services/dynamic-sidecar/tests/unit/test_modules_outputs_event_filter.py

```diff
@@ -0,0 +1,235 @@
+
+# FIXTURES
+
+
+@pytest.fixture
+def outputs_path(tmp_path: Path) -> Path:
+    return tmp_path
+
+
+@pytest.fixture
+def port_key_1() -> str:
+    return "port_1"
+
+
+@pytest.fixture
+def port_keys(outputs_path: Path, port_key_1: str) -> list[str]:
+    (outputs_path / port_key_1).mkdir(parents=True, exist_ok=True)
+    return [port_key_1]
+
+
+@pytest.fixture
+async def outputs_context(outputs_path: Path, port_keys: list[str]) -> OutputsContext:
+    outputs_context = OutputsContext(outputs_path)
+    await outputs_context.set_file_type_port_keys(port_keys)
+    return outputs_context
+
+
+@pytest.fixture
+async def outputs_manager(
+    outputs_context: OutputsContext,
+) -> AsyncIterator[OutputsManager]:
+    outputs_manager = OutputsManager(
+        outputs_context=outputs_context, io_log_redirect_cb=None
+    )
+    await outputs_manager.start()
+    yield outputs_manager
+    await outputs_manager.shutdown()
+
+
+@pytest.fixture
+def mocked_port_key_content_changed(
+    mocker: MockerFixture, outputs_manager: OutputsManager
+) -> Iterator[AsyncMock]:
+    async def _mock_upload_outputs(*args, **kwargs) -> None:
+        pass
+
+    yield mocker.patch.object(
+        outputs_manager, "port_key_content_changed", side_effect=_mock_upload_outputs
+    )
+
+
+@pytest.fixture
+def mock_delay_policy() -> BaseDelayPolicy:
+    VERY_FAST_BASE = 0.001
+    SLOW_LONG_INTERVAL = VERY_FAST_BASE * 100
+
+    # pylint:disable=no-self-use
+    class FastPolicy(BaseDelayPolicy):
+        def get_min_interval(self) -> NonNegativeFloat:
+            return VERY_FAST_BASE
+
+        def get_wait_interval(self, dir_size: NonNegativeInt) -> NonNegativeFloat:
+            return VERY_FAST_BASE if dir_size == 0 else SLOW_LONG_INTERVAL
+
+    return FastPolicy()
+
+
+@pytest.fixture
+def mock_get_dir_size(mocker: MockerFixture) -> Iterator[AsyncMock]:
+    yield mocker.patch(
+        "simcore_service_dynamic_sidecar.modules.outputs._event_filter.get_dir_size",
+        return_value=1,
+    )
+
+
+@pytest.fixture
+async def event_filter(
+    outputs_manager: OutputsManager, mock_delay_policy: BaseDelayPolicy
+) -> AsyncIterator[EventFilter]:
+    event_filter = EventFilter(
+        outputs_manager=outputs_manager, delay_policy=mock_delay_policy
+    )
+    await event_filter.start()
+    yield event_filter
+    await event_filter.shutdown()
+
+
+# UTILS
+
+
+async def _wait_for_event_to_trigger(event_filter: EventFilter) -> None:
+    await asyncio.sleep(event_filter.delay_policy.get_min_interval() * 2)
+
+
+async def _wait_for_event_to_trigger_big_directory(event_filter: EventFilter) -> None:
+    await asyncio.sleep(event_filter.delay_policy.get_wait_interval(1) * 2)
+
+
+# TESTS
+
+
+async def test_event_triggers_once(
+    event_filter: EventFilter,
+    port_key_1: str,
+    mocked_port_key_content_changed: AsyncMock,
+):
+    # event triggers once
+    await event_filter.enqueue(port_key_1)
+    await _wait_for_event_to_trigger(event_filter)
```

Comment:
why not instead of waiting, you just check ``mocked_port_key_content_changed.call_count == 1`` and retry on failure with a timeout (the value u are using)?

Reply:
Because the timing is what I want to test here, not the fact that it eventually happens.